src/freenet/client/filter/GIFFilter.java

```diff
@@ -30,22 +35,452 @@
+				throwDataError(l10n("invalidHeaderTitle"), l10n("invalidHeader"));
+			}
+			writeScreenDescriptor();
+			final boolean hasGlobalColorMap = (screenFlags & 0x80) == 0x80;
+			if (hasGlobalColorMap) {
+				copy(3 * screenColors);
+			}
+			filterData();
+		}
+
+		/** Reads the screen descriptor from the input and parses it. */
+		private void readScreenDescriptor() throws IOException, DataFilterException {
+			screenWidth = readShort();
+			screenHeight = readShort();
+			screenFlags = readByte();
+			screenBackgroundColor = readByte();
+			screenAspectRatio = readByte();
+
+			final int bitsPerPixel = (screenFlags & 0x07) + 1;
+			screenColors = 1 << bitsPerPixel;
+		}
+
+		/** Writes the previously parsed and validated screen descriptor to the output. */
+		private void writeScreenDescriptor() throws IOException {
+			writeShort(screenWidth);
+			writeShort(screenHeight);
+			writeByte(screenFlags);
+			writeByte(screenBackgroundColor);
+			writeByte(screenAspectRatio);
+		}
+
+		/** Looks for data blocks and filters them according to their type. */
+		private void filterData() throws IOException, DataFilterException {
+			boolean imageSeen = false;
+			boolean terminated = false;
+			int lastByte;
+			while (!terminated && (lastByte = input.read()) != -1) {
+				switch(lastByte) {
+					case IMAGE_SEPARATOR:
+						imageSeen |= filterImage();
+						break;
+					case GIF_TERMINATOR:
+						terminated |= imageSeen;
+						break;
+					case EXTENSION_INTRODUCER:
+						filterExtensionBlock();
+						break;
+					default:
+						// The specification expects us to skip other data; we can simply omit it.
+				}
+			}
+			if (!imageSeen) {
+				throwDataError(l10n("noDataTitle"), l10n("noData"));
+			}
+			if (!terminated) {
+				throwDataError(l10n("unterminatedGifTitle"), l10n("unterminatedGif"));
+			}
+			// There may still be trailing data at this point; we can simply omit it.
+			writeByte(GIF_TERMINATOR);
+		}
+
+		/** Filters a render block. Actual LZW data is *not* checked. */
+		private boolean filterImage() throws IOException, DataFilterException {
+			final int imageLeft = readShort();
+			final int imageTop = readShort();
+			final int imageWidth = readShort();
+			final int imageHeight = readShort();
+			final int imageFlags = readByte();
+			final boolean hasLocalColorMap = (imageFlags & 0x80) == 0x80;
+			final byte[] localColorMap;
+			if (hasLocalColorMap) {
+				final int bitsPerPixel = (imageFlags & 0x07) + 1;
+				final int imageColors = 1 << bitsPerPixel;
+				localColorMap = readBytes(3 * imageColors);
+			} else {
+				localColorMap = new byte[0];
+			}
+			final int lzwCodeSize = readByte();
+			if (imageLeft + imageWidth > screenWidth || imageTop + imageHeight > screenHeight ||
+					lzwCodeSize < 2 || lzwCodeSize >= 12 || !validateImageFlags(imageFlags)) {
+				foundImageData(false);
+				skipSubBlocks();
+				return false;
+			} else {
+				foundImageData(true);
+				writeByte(0x2c);
```

Comment:
What’s the function of this `0x2c` here? A constant with a descriptive name is preferrable.